app/gui/src/project-view/components/ComponentBrowser/ComponentList.vue

```diff
@@ -1,262 +1,173 @@
 <script setup lang="ts">
 import { makeComponentList, type Component } from '@/components/ComponentBrowser/component'
-import { Filtering } from '@/components/ComponentBrowser/filtering'
-import { useScrolling } from '@/components/ComponentBrowser/scrolling'
-import SvgIcon from '@/components/SvgIcon.vue'
-import { useApproach } from '@/composables/animation'
-import { useResizeObserver } from '@/composables/events'
+import ComponentEntry from '@/components/ComponentBrowser/ComponentEntry.vue'
+import type { Filtering } from '@/components/ComponentBrowser/filtering'
+import LazyList from '@/components/LazyList.vue'
 import { groupColorStyle } from '@/composables/nodeColors'
 import { useSuggestionDbStore } from '@/stores/suggestionDatabase'
 import { tryGetIndex } from '@/util/data/array'
-import { allRanges } from '@/util/data/range'
-import { computed, ref, watch } from 'vue'
+import { computed, ref, toRef, watch, watchEffect } from 'vue'
+import type { ComponentExposed } from 'vue-component-type-helpers'
 
-const ITEM_SIZE = 32
+const ITEM_SIZE = 24
+const SCROLL_TO_SELECTION_MARGIN = ITEM_SIZE / 2
+const MOUSE_SELECTION_DEBOUNCE = 200
 
 const props = defineProps<{
   filtering: Filtering
-  autoSelectFirstComponent: boolean
 }>()
 const emit = defineEmits<{
   acceptSuggestion: [suggestion: Component]
   'update:selectedComponent': [selected: Component | null]
 }>()
 
-const suggestionDbStore = useSuggestionDbStore()
+const root = ref<HTMLElement>()
+const groupsPanel = ref<ComponentExposed<typeof LazyList>>()
+const componentsPanel = ref<ComponentExposed<typeof LazyList>>()
+const panels = { groupsPanel, componentsPanel }
+export type ComponentListPanel = keyof typeof panels
 
-// === Components List and Positions ===
+const selectedGroupIndex = ref<number | null>(0)
+const selectedComponentIndex = ref<number | null>(0)
+const focusedPanel = ref<ComponentListPanel>('componentsPanel')
 
-const components = computed(() => makeComponentList(suggestionDbStore.entries, props.filtering))
+// When filtering updates, force componentList to be focused.
+watch(toRef(props, 'filtering'), () => (selectedComponentIndex.value = 0))
```

Comment:
It looks like this can focus a list entry even when `focusedPanel` is `groupsPanel`, if we receive a late filter update from the backend?

I see `selectedComponentIndex` is set to null/non-null from different places in order to focus/unfocus--I think this could be done more declaratively with a ref that only holds a user-selected index, and a `computed` value `displayedSelectedIndex` along the lines of `focusedPanel.value === 'groupsPanel' ? null : selectedComponentIndex.value ?? (currentComponents.value.length > 0 ? null)`